File: examples/submit_argo_workflow.py
#!/usr/bin/env python3
"""
Argo Workflow Submission Example (Python)

This script demonstrates how to submit a workflow to an Argo Workflows server.

Usage:
    python examples/submit_argo_workflow.py

Environment Variables:
    ARGO_SERVER - Argo Workflows server URL (default: https://localhost:2746)
    ARGO_TOKEN - Bearer token for authentication
    ARGO_NAMESPACE - Kubernetes namespace (default: default)
"""

import os
import sys
import json
import requests

# Disable SSL warnings for self-signed certificates
import urllib3
urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

# Configuration
ARGO_SERVER = os.environ.get('ARGO_SERVER', 'https://localhost:2746')
ARGO_TOKEN = os.environ.get('ARGO_TOKEN')
ARGO_NAMESPACE = os.environ.get('ARGO_NAMESPACE', 'default')

# Validate required environment variables
if not ARGO_TOKEN or ARGO_TOKEN == 'YOUR_TOKEN_HERE':
    print('Error: ARGO_TOKEN environment variable is required.', file=sys.stderr)
    print('Usage: ARGO_TOKEN=your_token python examples/submit_argo_workflow.py', file=sys.stderr)
    sys.exit(1)

# The workflow definition to submit
workflow = {
    'workflow': {
        'apiVersion': 'argoproj.io/v1alpha1',
        'kind': 'Workflow',
        'metadata': {
            'generateName': 'hello-argo-'
        },
        'spec': {
            'entrypoint': 'main',
            'templates': [
                {
                    'name': 'main',
                    'container': {
                        'image': 'alpine:latest',
                        'command': ['sh', '-c'],
                        'args': ['echo Hello, Argo!']
                    }
                }
            ]
        }
    }
}


def submit_workflow():
    """
    Submit a workflow to the Argo Workflows server.
    
    Returns:
        dict: The created workflow object
    """
    url = f'{ARGO_SERVER}/api/v1/workflows/{ARGO_NAMESPACE}'
    
    print(f'Submitting workflow to {url}...')
    
    try:
        response = requests.post(
            url,
            json=workflow,
            headers={
                'Content-Type': 'application/json',
                'Authorization': f'Bearer {ARGO_TOKEN}'
            },
            verify=False  # Skip SSL certificate verification
        )
        
        response.raise_for_status()
        
        data = response.json()
        print('\n✓ Workflow submitted successfully!')
        print(f"  Workflow name: {data['metadata']['name']}")
        print(f"  Workflow UID: {data['metadata']['uid']}")
        print(f"  Namespace: {data['metadata']['namespace']}")
        print(f"  Status: {data.get('status', {}).get('phase', 'Pending')}")
        
        return data
    except requests.exceptions.HTTPError as error:
        print('\n✗ Error submitting workflow:')
        print(f'  Status: {error.response.status_code}')
        try:
            error_data = error.response.json()
            print(f"  Message: {error_data.get('message', error.response.text)}")
            if 'details' in error_data:
                print(f"  Details: {json.dumps(error_data['details'], indent=2)}")
        except json.JSONDecodeError:
            print(f'  Message: {error.response.text}')
        raise
    except requests.exceptions.RequestException as error:
        print('\n✗ Error submitting workflow:')
        print(f'  Error: {error}')
        raise


if __name__ == '__main__':
    try:
        submit_workflow()
        print('\nDone!')
        sys.exit(0)
    except Exception:
        print('\nFailed!')
        sys.exit(1)

File: examples/submitArgoWorkflow.js
/**
 * Argo Workflow Submission Example
 * 
 * This script demonstrates how to submit a workflow to an Argo Workflows server.
 * 
 * Usage:
 *   node examples/submitArgoWorkflow.js
 * 
 * Environment Variables:
 *   ARGO_SERVER - Argo Workflows server URL (default: https://localhost:2746)
 *   ARGO_TOKEN - Bearer token for authentication
 *   ARGO_NAMESPACE - Kubernetes namespace (default: default)
 */

const axios = require('axios');
const https = require('https');

// Configuration
const ARGO_SERVER = process.env.ARGO_SERVER || 'https://localhost:2746';
const ARGO_TOKEN = process.env.ARGO_TOKEN;
const ARGO_NAMESPACE = process.env.ARGO_NAMESPACE || 'default';

// Validate required environment variables
if (!ARGO_TOKEN || ARGO_TOKEN === 'YOUR_TOKEN_HERE') {
  console.error('Error: ARGO_TOKEN environment variable is required.');
  console.error('Usage: ARGO_TOKEN=your_token node examples/submitArgoWorkflow.js');
  process.exit(1);
}

// The workflow definition to submit
const workflow = {
  workflow: {
    apiVersion: 'argoproj.io/v1alpha1',
    kind: 'Workflow',
    metadata: {
      generateName: 'hello-argo-'
    },
    spec: {
      entrypoint: 'main',
      templates: [
        {
          name: 'main',
          container: {
            image: 'alpine:latest',
            command: ['sh', '-c'],
            args: ['echo Hello, Argo!']
          }
        }
      ]
    }
  }
};

// Create an HTTPS agent that ignores certificate validation
// WARNING: This is insecure and should only be used in development
// In production, use proper SSL certificates and remove this configuration
const agent = new https.Agent({
  rejectUnauthorized: false
});

/**
 * Submit a workflow to the Argo Workflows server
 * @returns {Promise<Object>} The created workflow object
 */
async function submitWorkflow() {
  const url = `${ARGO_SERVER}/api/v1/workflows/${ARGO_NAMESPACE}`;
  
  console.log(`Submitting workflow to ${url}...`);
  
  try {
    const response = await axios.post(url, workflow, {
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${ARGO_TOKEN}`
      },
      httpsAgent: agent
    });

    console.log('\n✓ Workflow submitted successfully!');
    console.log(`  Workflow name: ${response.data.metadata.name}`);
    console.log(`  Workflow UID: ${response.data.metadata.uid}`);
    console.log(`  Namespace: ${response.data.metadata.namespace}`);
    console.log(`  Status: ${response.data.status?.phase || 'Pending'}`);
    
    return response.data;
  } catch (error) {
    console.error('\n✗ Error submitting workflow:');
    if (error.response) {
      // The request was made and the server responded with a status code
      // that falls out of the range of 2xx
      console.error(`  Status: ${error.response.status}`);
      console.error(`  Message: ${error.response.data?.message || error.response.statusText}`);
      if (error.response.data?.details) {
        console.error(`  Details: ${JSON.stringify(error.response.data.details, null, 2)}`);
      }
    } else if (error.request) {
      // The request was made but no response was received
      console.error('  No response received from server');
      console.error(`  Error: ${error.message}`);
    } else {
      // Something happened in setting up the request that triggered an Error
      console.error(`  Error: ${error.message}`);
    }
    throw error;
  }
}

// Execute the function if this script is run directly
if (require.main === module) {
  submitWorkflow()
    .then(() => {
      console.log('\nDone!');
      process.exit(0);
    })
    .catch(() => {
      console.log('\nFailed!');
      process.exit(1);
    });
}

module.exports = { submitWorkflow };
